### ISSUE TITLE
cloud-init not invoked on openstack using config drive

### ISSUE THREAD

**Issue by crawford** [coreos/bugs#735]
<a href="https://github.com/tryggth"><img src="https://avatars.githubusercontent.com/u/656385?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [tryggth](https://github.com/tryggth)**
_Friday May 30, 2014 at 21:13 GMT_
_Originally opened as https://github.com/coreos/coreos-cloudinit/issues/131_

----

Though /media/configdrive is mounted with the correct contents, cloud-config yaml file configuration is not applied.

Information:
coreos openstack image: #317
```
host-96-119-0-48 log # df
Filesystem     1K-blocks   Used Available Use% Mounted on
rootfs          39517164  16868  39496160   1% /
devtmpfs         2015684      0   2015684   0% /dev
tmpfs            2026356      0   2026356   0% /dev/shm
tmpfs            2026356    236   2026120   1% /run
tmpfs            2026356      0   2026356   0% /sys/fs/cgroup
/dev/vda9       39517164  16868  39496160   1% /
/dev/vda3        1032088 293592    686068  30% /usr
tmpfs            2026356      0   2026356   0% /media
tmpfs            2026356      0   2026356   0% /tmp
/dev/vda6         110576     68    101336   1% /usr/share/oem
/dev/sr0             422    422         0 100% /media/configdrive
```

```
host-96-119-0-48 log # ls -R /media/configdrive
/media/configdrive:
ec2  openstack

/media/configdrive/ec2:
2009-04-04  latest

/media/configdrive/ec2/2009-04-04:
meta-data.json  user-data

/media/configdrive/ec2/latest:
meta-data.json  user-data

/media/configdrive/openstack:
2012-08-10  2013-04-04  2013-10-17  latest

/media/configdrive/openstack/2012-08-10:
meta_data.json  user_data

/media/configdrive/openstack/2013-04-04:
meta_data.json  user_data

/media/configdrive/openstack/2013-10-17:
meta_data.json  user_data  vendor_data.json

/media/configdrive/openstack/latest:
meta_data.json  user_data  vendor_data.json
```

```
host-96-119-0-48 log # cat /media/configdrive/openstack/latest/user_data
#cloud-config
# this is a sample cloud-config which is known working on OpenStack
# the definition starts etcd & fleet, and creates a user 'core' with
# password 'coreos'

users:
  - name: core
    passwd: $1$WEcH/2xY$H6XiwyUpcnZ5Mvcu0YqQD0
    ssh-authorized-keys:
      - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCobfm/jGSrZ4aABl+JAXwTYZjWAklBtFfr+TiVJCGXqt2Rs1L8B6K36i0vWyOqQWfrkveiF8/DderPgOb4BvGNyP+IROXPfSJqi0lN6zlfQSswKdXPzefohn5uO2JJtdEfftuR3Lxc63ZBLh1fg6zCBaqOrEilcRIU1kAitQec8VoMP9nqHwRcBiW5YlodNoubDKz6Ic4Ps3NH0UHHivPHUxCcZ2YBP0AV/GJIQQGWkDXYcRGNZa3cci1I/3icV/rnThn2wWWQIPMr3yucHxT/7Y7zTi0O1BJJ5hYz1hJbqtqORzeHmcrD+KIg4arBQwPOXE7Tb5Wyn0D/vXSgKnvl Generated by Nova
```

journal logs at:
https://drive.google.com/folderview?id=0B2FgUvsnYYRbYVVvLUN4SF9Xa2M&usp=sharing
https://docs.google.com/document/d/1pqNlne_fIu3_CVLPf1_YdRJx3eLwYi-PN-OeQBAnxl8/edit?usp=sharing
https://drive.google.com/file/d/0B2FgUvsnYYRbc1RnR3JweEhjNEE/edit?usp=sharing

**Comment by crawford** [comment]
<a href="https://github.com/jonboulle"><img src="https://avatars.githubusercontent.com/u/2601015?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jonboulle](https://github.com/jonboulle)**
_Friday Jun 06, 2014 at 20:09 GMT_

----

What do these show?
```
systemctl status user-configdrive.path
systemctl status user-configdrive.service
```

**Comment by crawford** [comment]
<a href="https://github.com/tryggth"><img src="https://avatars.githubusercontent.com/u/656385?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tryggth](https://github.com/tryggth)**
_Friday Jun 06, 2014 at 20:43 GMT_

----

core@host-96-119-0-48 ~ $ systemctl status user-configdrive.path
â user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static)
   Active: active (running) since Thu 2014-06-05 00:39:29 UTC; 1 day 20h ago

Jun 05 00:39:29 localhost systemd[1]: Starting Watch for a cloud-config at /media/configdrive.
Jun 05 00:39:29 localhost systemd[1]: Started Watch for a cloud-config at /media/configdrive.
core@host-96-119-0-48 ~ $ systemctl status user-configdrive.service
â user-configdrive.service - Load cloud-config from /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.service; static)
   Active: inactive (dead)

**Comment by crawford** [comment]
<a href="https://github.com/tryggth"><img src="https://avatars.githubusercontent.com/u/656385?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tryggth](https://github.com/tryggth)**
_Monday Jun 09, 2014 at 13:59 GMT_

----

Anything else you need?

**Comment by crawford** [comment]
<a href="https://github.com/crawford"><img src="https://avatars.githubusercontent.com/u/368848?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [crawford](https://github.com/crawford)**
_Monday Jun 09, 2014 at 16:45 GMT_

----

Can you tell me output from `systemctl list-jobs`?

**Comment by crawford** [comment]
<a href="https://github.com/tryggth"><img src="https://avatars.githubusercontent.com/u/656385?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tryggth](https://github.com/tryggth)**
_Monday Jun 09, 2014 at 17:29 GMT_

----

core@host-96-119-0-48 ~ $ systemctl list-jobs
JOB UNIT                                                       TYPE  STATE
103 multi-user.target                                          start waiting
167 user-config.target                                         start waiting
168 system-config.target                                       start waiting
169 coreos-setup-environment.service                           start running
170 system-cloudinit@usr-share-oem-cloud\x2dconfig.yml.service start running
175 user-cloudinit-proc-cmdline.service                        start waiting
350 user-configdrive.service                                   start waiting
556 ec2-ssh-key.service                                        start running

8 jobs listed.
core@host-96-119-0-48 ~ $

**Comment by crawford** [comment]
<a href="https://github.com/crawford"><img src="https://avatars.githubusercontent.com/u/368848?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [crawford](https://github.com/crawford)**
_Monday Jun 09, 2014 at 17:58 GMT_

----

Ok, this is the same issue as #135. The underlying issue is that the openstack images are trying to connect to the metadata service in coreos-setup-environment.service (which it never finds in your case). This causes a series of services to block (the 8 jobs in your last printout), waiting on the connection to the metadata service.

**Comment by crawford** [comment]
<a href="https://github.com/cgroschupp"><img src="https://avatars.githubusercontent.com/u/6190136?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [cgroschupp](https://github.com/cgroschupp)**
_Monday Jun 09, 2014 at 21:23 GMT_

----

I think this a another problem than #135.
I do not know how DirectoryNotEmpty in systemd works (probably via a notify), but I think the order from the mount and check whether files exist is wrong.

After unmounting /media/configdrive/ and create a file in /media/configdrive/, the unit is running.

● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static)
   Active: active (waiting) since Mon 2014-06-09 19:00:41 UTC; 2min 16s ago

Jun 09 19:00:41 node002 systemd[1]: Starting Watch for a cloud-config at /media/configdrive.
Jun 09 19:00:41 node002 systemd[1]: Started Watch for a cloud-config at /media/configdrive.

   
sudo umount /media/configdrive/
sudo touch /media/configdrive/test

● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static)
   Active: active (running) since Mon 2014-06-09 19:00:41 UTC; 2min 27s ago

Jun 09 19:00:41 node002 systemd[1]: Starting Watch for a cloud-config at /media/configdrive.
Jun 09 19:00:41 node002 systemd[1]: Started Watch for a cloud-config at /media/configdrive.

**Comment by crawford** [comment]
<a href="https://github.com/jakedahn"><img src="https://avatars.githubusercontent.com/u/8596?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jakedahn](https://github.com/jakedahn)**
_Thursday Oct 02, 2014 at 06:47 GMT_

----

/cc @crawford 

To start kicking the tires I tried to use the openstack friendly image that I found here http://alpha.release.core-os.net/amd64-usr/current/coreos_production_openstack_image.img.bz2

When I tried to use the `$private_ipv4`/`$public_ipv4` variables in the cloud-init config as specified in the docs here: https://coreos.com/docs/cluster-management/setup/cloudinit-cloud-config/ -- I noticed that the variables were never properly being replaced (the etcd config was ommiting an ip address and just printing out the port number). I started digging around and I think it may be related to the failure of a cloud-configdrive units that failed on boot. I'm able to get info from the metadata service on `169.254.169.254`, and I'm also able to browse around and see data in `/media/configdrive`.

Not sure if this is the same bug, if its a new bug, or if I'm just doing something silly. Below is my cloud-config, and output about failed units.

```yaml
#cloud-config

coreos:
  etcd:
    discovery: https://discovery.etcd.io/magichash
    addr: $private_ipv4:4001
    peer-addr: $private_ipv4:7001
  units:
    - name: etcd.service
      command: start
    - name: fleet.service
      command: start
```

```
$ systemctl status user-configdrive.path
● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static)
   Active: failed (Result: resources)

Oct 02 05:47:25 localhost systemd[1]: Starting Watch for a cloud-config at /media/configdrive.
Oct 02 05:47:25 localhost systemd[1]: Started Watch for a cloud-config at /media/configdrive.
Oct 02 06:11:40 core0.novalocal systemd[1]: Starting Watch for a cloud-config at /media/configdrive.
Oct 02 06:11:40 core0.novalocal systemd[1]: Failed to start Watch for a cloud-config at /media/configdrive.

$ systemctl status user-configdrive.service
● user-configdrive.service
   Loaded: masked (/dev/null)
   Active: inactive (dead)

Oct 02 05:47:25 core0 systemd[1]: Failed to start user-configdrive.service.
```

**Comment by crawford** [comment]
<a href="https://github.com/crawford"><img src="https://avatars.githubusercontent.com/u/368848?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [crawford](https://github.com/crawford)**
_Thursday Oct 09, 2014 at 16:53 GMT_

----

@jakedahn This looks like a different bug. Can you open a new issue?

**Comment by crawford** [comment]
<a href="https://github.com/dsampath"><img src="https://avatars.githubusercontent.com/u/1567374?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [dsampath](https://github.com/dsampath)**
_Thursday Oct 09, 2014 at 21:22 GMT_

----

I see a similar error - could you point me to the issue if it's been created?

[update]: I am using openstack icehouse (on CentOs) to deploy this. I got the image off of [coreOs site](https://coreos.com/docs/running-coreos/platforms/openstack/) and tried to deploy it in my cluster using the guide. 

* I see that the ip addresses in the meta-data hasn't been updated. 
* I also noticed the error message below when I sshed in to the box. 
```
CoreOS (alpha)
Failed Units: 1
  oem-cloudinit.service
```


```
core@coreos-143fe30e-87f5-4e7b-bdda-f173b578794c ~ $ fleetctl list-machines
2014/10/09 21:13:19 INFO client.go:278: Failed getting response from http://127.0.0.1:4001/: dial tcp 127.0.0.1:4001: connection refused
2014/10/09 21:13:19 ERROR client.go:200: Unable to get result for {Get /_coreos.com/fleet/machines}, retrying in 100ms
2014/10/09 21:13:19 INFO client.go:278: Failed getting response from http://127.0.0.1:4001/: dial tcp 127.0.0.1:4001: connection refused
2014/10/09 21:13:19 ERROR client.go:200: Unable to get result for {Get /_coreos.com/fleet/machines}, retrying in 200ms
2014/10/09 21:13:19 INFO client.go:278: Failed getting response from http://127.0.0.1:4001/: dial tcp 127.0.0.1:4001: connection refused
2014/10/09 21:13:19 ERROR client.go:200: Unable to get result for {Get /_coreos.com/fleet/machines}, retrying in 400ms
2014/10/09 21:13:19 INFO client.go:278: Failed getting response from http://127.0.0.1:4001/: dial tcp 127.0.0.1:4001: connection refused
2014/10/09 21:13:19 ERROR client.go:200: Unable to get result for {Get /_coreos.com/fleet/machines}, retrying in 800ms
2014/10/09 21:13:20 INFO client.go:278: Failed getting response from http://127.0.0.1:4001/: dial tcp 127.0.0.1:4001: connection refused
2014/10/09 21:13:20 ERROR client.go:200: Unable to get result for {Get /_coreos.com/fleet/machines}, retrying in 1s
^C

core@coreos-143fe30e-87f5-4e7b-bdda-f173b578794c ~ $ systemctl status user-configdrive.path
● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static)
   Active: failed (Result: resources)

Oct 09 21:09:01 localhost systemd[1]: Started Watch for a cloud-config at /media/configdrive.


core@coreos-143fe30e-87f5-4e7b-bdda-f173b578794c ~ $ systemctl status user-configdrive.service
● user-configdrive.service
   Loaded: masked (/dev/null)
   Active: inactive (dead)

Oct 09 21:09:01 localhost systemd[1]: Failed to start user-configdrive.service.


core@coreos-143fe30e-87f5-4e7b-bdda-f173b578794c ~ $ systemctl list-jobs
No jobs running.


core@coreos-143fe30e-87f5-4e7b-bdda-f173b578794c /media/configdrive/openstack/latest $ cat user_data 
#cloud-config

coreos:
  etcd:
    # generate a new token for each unique cluster from https://discovery.etcd.io/new
    discovery: https://discovery.etcd.io/<valid token here - removed for comment>
    # multi-region and multi-cloud deployments need to use $public_ipv4
    addr: $private_ipv4:4001
    peer-addr: $private_ipv4:7001
  units:
    - name: etcd.service
      command: start
    - name: fleet.service
      command: start
ssh_authorized_keys:
```

**Comment by crawford** [comment]
<a href="https://github.com/tvb"><img src="https://avatars.githubusercontent.com/u/1263414?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [tvb](https://github.com/tvb)**
_Thursday Nov 20, 2014 at 14:19 GMT_

----

Want to hook into this issue as well, having the exact same issue and output as explained by @dsampath

**Comment by matthughes** [comment]
Any update on this issue?  I seem to be getting something very similar.  My OS tenant doesn't have access to the metadata service, but I do have config drive running.  I don't actually use any environment variables from metadata; I rewrite the file before boot so I don't need to do any variable substitution.  But configdrive consistently fails to come up.

I've checked the user-data is valid via CoreOS's online validator.  

```
core@master ~ $ systemctl status user-configdrive.path
● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static; vendor preset: disabled)
   Active: failed (Result: resources)

Feb 19 22:50:40 localhost systemd[1]: Started Watch for a cloud-config at /media/configdrive.
Feb 19 22:50:40 localhost systemd[1]: user-configdrive.path: Failed to queue unit startup job: Unit user-configdrive.service is masked.
core@master ~ $ systemctl status user-configdrive.service
● user-configdrive.service
   Loaded: masked (/dev/null)
   Active: inactive (dead)

Feb 19 22:50:40 host-96-119-240-193 systemd[1]: Failed to start user-configdrive.service.
core@master ~ $ systemctl list-jobs
  JOB UNIT                   TYPE  STATE
25229 kube-apiserver.service start running

1 jobs listed.
core@master ~ $ cat /media/configdrive/openstack/latest/user_data
#cloud-config

---
write-files:
  - path: /etc/conf.d/nfs
    permissions: '0644'
    content: |
      OPTS_RPC_MOUNTD=""
  - path: /opt/bin/wupiao
    permissions: '0755'
    content: |
      #!/bin/bash
      # [w]ait [u]ntil [p]ort [i]s [a]ctually [o]pen
      [ -n "$1" ] && \
        until curl -o /dev/null -sIf http://1; do \
          sleep 1 && echo .;
        done;
      exit $?

hostname: master
coreos:
  etcd2:
    name: master
    listen-client-urls: http://0.0.0.0:2379,http://0.0.0.0:4001
    advertise-client-urls: http://96.119.240.193:2379,http://96.119.240.193:4001
    initial-cluster-token: k8s_etcd
    listen-peer-urls: http://96.119.240.193:2380,http://96.119.240.193:7001
    initial-advertise-peer-urls: http://96.119.240.193:2380
    initial-cluster: master=http://96.119.240.193:2380
    initial-cluster-state: new
  fleet:
    metadata: "role=master"
  units:
    - name: generate-serviceaccount-key.service
      command: start
      content: |
        [Unit]
        Description=Generate service-account key file

        [Service]
        ExecStartPre=-/usr/bin/mkdir -p /opt/bin
        ExecStart=/bin/openssl genrsa -out /opt/bin/kube-serviceaccount.key 2048 2>/dev/null
        RemainAfterExit=yes
        Type=oneshot
    - name: setup-network-environment.service
      command: start
      content: |
        [Unit]
        Description=Setup Network Environment
        Documentation=https://github.com/kelseyhightower/setup-network-environment
        Requires=network-online.target
        After=network-online.target

        [Service]
        ExecStartPre=-/usr/bin/mkdir -p /opt/bin
        ExecStartPre=/usr/bin/curl -L -o /opt/bin/setup-network-environment -z /opt/bin/setup-network-environment https://github.com/kelseyhightower/setup-network-environment/releases/download/v1.0.0/setup-network-environment
        ExecStartPre=/usr/bin/chmod +x /opt/bin/setup-network-environment
        ExecStart=/opt/bin/setup-network-environment
        RemainAfterExit=yes
        Type=oneshot
    - name: fleet.service
      command: start
    - name: flanneld.service
      command: start
      drop-ins:
        - name: 50-network-config.conf
          content: |
            [Unit]
            Requires=etcd2.service
            [Service]
            ExecStartPre=/usr/bin/etcdctl set /coreos.com/network/config '{"Network":"10.244.0.0/16", "Backend": {"Type": "vxlan"}}'
    - name: docker.service
      command: start
    - name: kube-apiserver.service
      command: start
      content: |
        [Unit]
        Description=Kubernetes API Server
        Documentation=https://github.com/GoogleCloudPlatform/kubernetes
        Requires=setup-network-environment.service etcd2.service generate-serviceaccount-key.service
        After=setup-network-environment.service etcd2.service generate-serviceaccount-key.service

        [Service]
        EnvironmentFile=/etc/network-environment
        ExecStartPre=-/usr/bin/mkdir -p /opt/bin
        ExecStartPre=/usr/bin/curl -L -o /opt/bin/kube-apiserver -z /opt/bin/kube-apiserver https://storage.googleapis.com/kubernetes-release/release/v1.0.3/bin/linux/amd64/kube-apiserver
        ExecStartPre=/usr/bin/chmod +x /opt/bin/kube-apiserver
        ExecStartPre=/opt/bin/wupiao 127.0.0.1:2379/v2/machines
        ExecStart=/opt/bin/kube-apiserver \
        --service-account-key-file=/opt/bin/kube-serviceaccount.key \
        --service-account-lookup=false \
        --admission-control=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota \
        --runtime-config=api/v1 \
        --allow-privileged=true \
        --insecure-bind-address=0.0.0.0 \
        --insecure-port=8080 \
        --kubelet-https=true \
        --secure-port=6443 \
        --service-cluster-ip-range=10.100.0.0/16 \
        --etcd-servers=http://127.0.0.1:2379 \
        --public-address-override=96.119.240.193 \
        --logtostderr=true
        Restart=always
        RestartSec=10
    - name: kube-controller-manager.service
      command: start
      content: |
        [Unit]
        Description=Kubernetes Controller Manager
        Documentation=https://github.com/GoogleCloudPlatform/kubernetes
        Requires=kube-apiserver.service
        After=kube-apiserver.service

        [Service]
        ExecStartPre=/usr/bin/curl -L -o /opt/bin/kube-controller-manager -z /opt/bin/kube-controller-manager https://storage.googleapis.com/kubernetes-release/release/v1.0.3/bin/linux/amd64/kube-controller-manager
        ExecStartPre=/usr/bin/chmod +x /opt/bin/kube-controller-manager
        ExecStart=/opt/bin/kube-controller-manager \
        --service-account-private-key-file=/opt/bin/kube-serviceaccount.key \
        --master=127.0.0.1:8080 \
        --logtostderr=true
        Restart=always
        RestartSec=10
    - name: kube-scheduler.service
      command: start
      content: |
        [Unit]
        Description=Kubernetes Scheduler
        Documentation=https://github.com/GoogleCloudPlatform/kubernetes
        Requires=kube-apiserver.service
        After=kube-apiserver.service

        [Service]
        ExecStartPre=/usr/bin/curl -L -o /opt/bin/kube-scheduler -z /opt/bin/kube-scheduler https://storage.googleapis.com/kubernetes-release/release/v1.0.3/bin/linux/amd64/kube-scheduler
        ExecStartPre=/usr/bin/chmod +x /opt/bin/kube-scheduler
        ExecStart=/opt/bin/kube-scheduler --master=127.0.0.1:8080
        Restart=always
        RestartSec=10
  update:
    group: alpha
    reboot-strategy: off
core@master ~ $
```

**Comment by matthughes** [comment]
I'm pretty new to CoreOS, but what does masking mean in the context of a service?  https://coreos.com/os/docs/latest/other-settings.html#mask-a-systemd-unit talks about masking, but my grub.cfg doesn't mask any services:

```
core@master ~ $ cat /usr/share/oem/grub.cfg
# CoreOS GRUB settings for EC2

set oem_id="openstack"
```

**Comment by robszumski** [comment]
@matthughes I'm not sure why your unit is failing due to lack of resources:

```
core@master ~ $ systemctl status user-configdrive.path
● user-configdrive.path - Watch for a cloud-config at /media/configdrive
   Loaded: loaded (/usr/lib64/systemd/system/user-configdrive.path; static; vendor preset: disabled)
   Active: failed (Result: resources)
```

Which is causing the rest to fail:

```
Feb 19 22:50:40 localhost systemd[1]: user-configdrive.path: Failed to queue unit startup job: Unit user-configdrive.service is masked.
```

@matthughes Masking prevents a service from running. Certain images of CoreOS mask these config related services to make sure that only the appropriate ones run.
